src/renderers/shared/fiber/ReactFiberScheduler.js

```diff
@@ -523,171 +562,136 @@ module.exports = function<T, P, I, TI, C>(config : HostConfig<T, P, I, TI, C>) {
+    if (!nextUnitOfWork && firstUncaughtError) {
+      let e = firstUncaughtError;
+      firstUncaughtError = null;
+      throw e;
     }
   }
 
-  function handleErrors() : Error | null {
-    if (activeErrorBoundaries) {
-      throw new Error('Already handling errors');
-    }
-
-    // Start tracking active boundaries.
-    activeErrorBoundaries = new Set();
-    // If we find unhandled errors, we'll only remember the first one.
-    let firstUncaughtError = null;
-
-    // All work created by error boundaries should have Task priority
-    // so that it finishes before this function exits.
-    const previousPriorityContext = priorityContext;
-    priorityContext = TaskPriority;
+  function captureError(failedWork : Fiber | null, error : Error, isUnmounting : boolean) : Fiber | null {
+    // It is no longer valid because we exited the user code.
+    ReactCurrentOwner.current = null;
+    // It is no longer valid because this unit of work failed.
+    nextUnitOfWork = null;
 
-    // Keep looping until there are no more trapped errors, or until we find
-    // an unhandled error.
-    while (nextTrappedErrors && nextTrappedErrors.length && !firstUncaughtError) {
-      // First, find all error boundaries and notify them about errors.
-      while (nextTrappedErrors && nextTrappedErrors.length) {
-        const trappedError = nextTrappedErrors.shift();
-        const boundary = trappedError.boundary;
-        const error = trappedError.error;
-        const root = trappedError.root;
-        if (!boundary) {
-          firstUncaughtError = firstUncaughtError || error;
-          if (root && root.current) {
-            // Unschedule this particular root since it fataled and we can't do
-            // more work on it. This lets us continue working on other roots
-            // even if one of them fails before rethrowing the error.
-            root.current.pendingWorkPriority = NoWork;
-          } else {
-            // Normally we should know which root caused the error, so it is
-            // unusual if we end up here. Since we assume this function always
-            // unschedules failed roots, our only resort is to completely
-            // unschedule all roots. Otherwise we may get into an infinite loop
-            // trying to resume work and finding the failing but unknown root again.
-            nextScheduledRoot = null;
-            lastScheduledRoot = null;
+    // Ignore this error if it's the result of unmounting a failed boundary
+    if (failedWork &&
+        isUnmounting &&
+        capturedErrors &&
+        capturedErrors.has(failedWork)) {
+      return null;
+    }
+
+    // Search for the nearest error boundary.
+    let boundary : Fiber | null = null;
+    if (failedWork) {
+      // Host containers are a special case. If the failed work itself is a host
+      // container, then it acts as its own boundary. In all other cases, we
+      // ignore the work itself and only search through the parents.
```

Comment:
If so, I'm not sure if we need to bother with this special case much since it happens at the end after all other work so there's no interleaving issue. Maybe for batching across roots but I'm not sure if this solution is sufficient in that case yet anyway.